src/io.h

```diff
@@ -167,83 +167,363 @@ class FormatGRO {
+     * @param[in] offset  offset in nanometers to add to all coordinates upon conversion
+     */
+    void importCoordinates(const PointVector &coordinates, const Point &offset);
+    /**
+     * @brief Imports and converts atomic coordinates. Offset is added to all coordinates to account different
+     * coordinates' origin.
+     * @tparam begin_iterator
+     * @tparam end_iterator
+     * @param[in] begin  input iterator with coordinates in nanometers
+     * @param[in] end  input iterator's end
+     * @param[in] offset  offset in nanometers to add to all coordinates upon conversion
+     */
+    template <class begin_iterator, class end_iterator>
+    void importCoordinates(begin_iterator begin, end_iterator end, const Point &offset) const {
+        // setNumberOfAtoms(std::distance(begin, end));
+        // comparison of i is probably faster than prior call of std::distance
+        int i = 0;
+        for (auto coordinates_it = begin; coordinates_it != end; ++coordinates_it) {
+            if (i >= number_of_atoms) {
+                throw std::runtime_error("number of particles too high to be saved in the XTC frame");
+            }
+            Point pos = (*coordinates_it + offset) / 1.0_nm;
+            XTCVector xtc_pos = pos.cast<XTCFloat>();
+            std::copy(xtc_pos.data(), xtc_pos.data() + DIM, xtc_coordinates.get()[i++]);
+        }
+        if (i < number_of_atoms) {
+            throw std::runtime_error("number of particles too low to be saved in the XTC frame");
+        }
+    }
+    /**
+     * @brief Exports and converts step and timestamp.
+     * @param[out] step  frame step
+     * @param[out] time  frame timestamp
+     */
+    void exportTimestamp(int &step, float &time) const;
+    /**
+     * @brief Exports and converts simulation box dimensions.
+     * @param[out] box  simulation box dimensions in nanometers (xyz)
+     */
+    void exportBox(Point &box) const;
+    /**
+     * @brief Exports and converts atomic coordinates. Offset is subtracted from all coordinates to account different
+     * coordinates' origin.
+     * @param[out] coordinates  atomic coordinates in nanometers
+     * @param[in] offset  offset in nanometers to subtract from all coordinates upon conversion
+     */
+    void exportCoordinates(PointVector &coordinates, const Point &offset) const;
+    /**
+     * @brief Exports and converts atomic coordinates. Offset is subtracted to all coordinates to account different
+     * coordinates' origin.
+     * @tparam begin_iterator
+     * @tparam end_iterator
+     * @param[out] begin  output iterator with coordinates in nanometers
+     * @param[out] end  output iterator's end
+     * @param[in] offset  offset in nanometers to subtract from all coordinates upon conversion
+     */
+    template <class begin_iterator, class end_iterator>
+    void exportCoordinates(begin_iterator begin, end_iterator end, const Point &offset) const {
```

Comment:
use `std::transform` and `std::begin(xtc_coordinates.get())`. Throw can be moved outside loop.

Reply:
Does not look as straightforward with 2D C-arrays. Keeping the current code.